rfcs/0002-tracing-cache.md

```diff
@@ -0,0 +1,95 @@
+# RFC-0000: Template
+
+<dl>
+  <dt>Author</dt>
+  <dd>Zac Burns</dd>
+
+  <dt>RFC pull request</dt>
+  <dd><a href="URL">URL</a></dd>
+
+  <dt>Obsoletes (if applicable)</dt>
+  <dd>None</dd>
+
+  <dt>Date of submission</dt>
+  <dd>2019-12-13</dd>
+
+  <dt>Date of approval</dt>
+  <dd>YYYY-MM-DD</dd>
+
+  <dt>Approved by</dt>
+  <dd>First Person, Second Person</dd>
+</dl>
+
+## Summary
+
+This RFC proposes the creation of a local etherium tracing cache to speed up indexing of subgraphs which use callHandler.
+
+## Motivation
+
+If a subgraph has callHandlers specified, it is necessary to extract calls from a trace of each block that a graph node indexes. It is expensive to acquire and process traces from etherium nodes, both in money and time.
+
+A production graph node may index multiple subgraphs and repeatedly access the same traces many times - especially for recent blocks. Having that data on hand in a pre-processed and easily filtered format may substantially reduce the cost of acquiring and converting trace data into triggers relevant to a subgraph.
+
+## Urgency
+
+None
+
+## Terminology
+
+None.
+
+## Detailed Design
+
+The design comes in 2 parts. The caching hierarchy, and the easily filtered data format.
+
+### Cache Hierarchy
+In `etherium_adapter.rs` is a function `traces`, which unconditionally makes a call to eth.web3. A hierarchical cache would be inserted here with 3 layers - In-Memory, On-Disk, and Fetch.
+
+#### In-Memory Layer
+At the top level of the cache is the in-memory cache. For a graph node which is indexing multiple subgraphs in parallel, it may be likely that they require the latest blocks simultaneously. By storing the buffer containing immutable trace data for a block in a rc::Weak, a low-risk, low-cost cache is introduced. The most recent trace, and all currently in-use traces would be maintained.
```

Comment:
The in-memory cache is clever, but the gain from it is less clear, so I'd suggest leaving it as a future extension.

Reply:
Yes. That the main purpose of the cache is for subgraph re-sync changes *everything*. The in memory cache was meant for requests for the same block across multiple subgraphs, but storing the trace for a whole block doesn't make sense anymore. Other parts of the RFC that don't make sense with this additional context include the futures re-use, the usefulness of zero-copy / skipping / lazy parse in the binary format, and... pretty much the whole RFC.

What's been nice about this RFC process is that although it's painful to re-write an RFC in response to missing information that pales in comparison to re-writing the code.

Reply:
That's (part of) the idea — thinking is cheaper than writing code 😄